src/Microsoft.ML.Tokenizers/Model/CodeGen.cs

```diff
@@ -0,0 +1,1723 @@
+
+                if (EndOfSentenceId.HasValue && id == EndOfSentenceId.Value)
+                {
+                    if (considerSpecialTokens)
+                    {
+                        AppendToBytesArray(EndOfSentenceToken!.AsSpan(), ref bytes, ref bytesIndex);
+                    }
+                    continue;
+                }
+
+                if (UnknownTokenId.HasValue && id == UnknownTokenId.Value)
+                {
+                    if (considerSpecialTokens)
+                    {
+                        AppendToBytesArray(UnknownToken!.AsSpan(), ref bytes, ref bytesIndex);
+                    }
+                    continue;
+                }
+
+                if (_addedTokensReverse is not null && _addedTokensReverse.TryGetValue(id, out string? addedToken))
+                {
+                    int bytesCountToEncode = Encoding.UTF8.GetMaxByteCount(addedToken.Length);
+                    if (bytes.Length - bytesIndex < bytesCountToEncode)
+                    {
+                        Helpers.ArrayPoolGrow(ref bytes, (bytes.Length + bytesCountToEncode) * 2);
+                    }
+
+                    bool removePrefixSpace = firstToken && hasPrefixSpace && addedToken.Length > 0 && addedToken[0] == ' ';
+                    bytesIndex += Helpers.GetUtf8Bytes(removePrefixSpace ? addedToken.AsSpan().Slice(1) : addedToken.AsSpan(), bytes.AsSpan().Slice(bytesIndex));
+                    firstToken = false;
+                    continue;
+                }
+
+                // vocabularies are stored in UTF-8 form with escaping the control characters.
+                // Need to convert the vocabulary to the original UTF-16 form.
+
+                if (MapIdToToken(id) is string s)
+                {
+                    ReadOnlySpan<char> span = firstToken && hasPrefixSpace && s.Length > 0 && s[0] == _transformedSpace ? s.AsSpan(1) : s.AsSpan();
+                    firstToken = false;
+                    AppendToBytesArray(span, ref bytes, ref bytesIndex);
+                }
+            }
+
+            string result = Encoding.UTF8.GetString(bytes, 0, bytesIndex);
+            ArrayPool<byte>.Shared.Return(bytes);
+            return result;
+        }
+
+        private void AppendToBytesArray(ReadOnlySpan<char> text, ref byte[] bytes, ref int bytesIndex)
+        {
+            for (int i = 0; i < text.Length; i++)
+            {
+                char c = text[i];
+                if ((uint)c < ByteToUnicodeEncoding.Instance.Count)
+                {
+                    if (bytesIndex >= bytes.Length)
+                    {
+                        Helpers.ArrayPoolGrow<byte>(ref bytes, bytes.Length * 2);
+                    }
+
+                    bytes[bytesIndex++] = (byte)ByteToUnicodeEncoding.Instance.UnicodeToByte[c];
+                    continue;
+                }
+
+                // rare cases
+                i += EncodeNonAnsiCodePointToUtf8(text, i, ref bytes, ref bytesIndex);
```

Comment:
Do we need this for .NET Core? We can't use Rune or some of the other built-in support?

Reply:
I fixed that and used Rune when running on net core.